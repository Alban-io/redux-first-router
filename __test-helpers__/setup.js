import createHistory from 'history/createMemoryHistory'
import connectRoutes from '../src/connectRoutes'


export default (path = '/', options = { title: 'title', location: 'location' }) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param',
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  return connectRoutes(history, routesMap, options)
}
