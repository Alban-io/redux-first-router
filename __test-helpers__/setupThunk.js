import { createStore, applyMiddleware, compose } from 'redux'
import createHistory from 'history/createMemoryHistory'
import connectTypes from '../src/connectTypes'


export default (path = '/', thunk) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second/:param', thunk },
    THIRD: { path: '/third/:param' },
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  const { middleware, enhancer, thunk: ssrThunk, reducer: locationReducer } = connectTypes(history, routesMap)

  const middlewares = applyMiddleware(middleware)

  const reducer = (state = {}, action = {}) => ({
    location: locationReducer(state.location, action),
  })

  const store = createStore(reducer, undefined, compose(enhancer, middlewares))

  return { store, thunk: ssrThunk }
}
