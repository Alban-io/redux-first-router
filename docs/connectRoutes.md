# connectRoutes

`connectRoutes` is the primary "work" you will do to get **Redux First Router** going. It's all about creating and maintaining
a pairing of *action types* and dynamic express style route paths. If you use our `<Link />` component and pass an *action* as
its `href` prop, you can change the URLs you use here any time without having to change your application code. 

In general, once you set this up, the only "work" you should expect to do in your component code is diligent use of the `<Link />` component
to specify what actions/URLs will be dispatched, and perhaps more importantly what markup will be written to the page. You can get the same
effect of the address bar changing just by dispatching actions bound as event handlers like you usually do, but that doesn't get you
`<a>` tags embedded in the page for search engines to pick up. 

To get the benefits of SEO, the changes to your workflow are minor. It's an inversion of control where you now specify actions in your components
rather than `mapDispatchToProps` handlers. Once you get used to it, it will take very little discpline to make this your new way of operating.

Lastly, it's not that you can't still use `mapDispatchToProps`--you just won't want to
in order to get SEO benefits for actions that you want to change the address bar (and allow the user to *back/next* through using the 
browser buttons). To learn more about how to create links, check out our [Link component](https://github.com/faceyspacey/redux-first-router-link).

Let's now examine our `connectRoutes` function more deeply:

## Signature
```javascript
connectRoutes(
  history: History,
  routesMap?: RoutesMap,
  options?: Options,
) : {
  middleware: Function,
  enhancer: Function,
  reducer: Function,
  thunk: Function,
}
```

The only parameter `connectRoutes` expects is the `history` object. You won't get much for that, but it won't break your redux store.
The second parameter is your all-important `routesMap`. And the third is a small set of `options` in a map that you can provide.

Let's dive into each.

## History
The `history` object is the return of the *history* package's `createBrowserHistory` or `createMemoryHistory` function:

```
import createHistory from 'history/createBrowserHistory'
const history = createHistory() // notice no parameters are needed in the browser
const { middleware, enhancer, reducer } = connectRoutes(history)

```
*or:*
```
import createHistory from 'history/createMemoryHistory'
const history = createHistory({ initialEntries: [request.path] })
const { middleware, enhancer, reducer } = connectRoutes(history)
```

See the widely used [history package](https://github.com/ReactTraining/history) on github. The idea is simply that you can use
both interchangeably depending on if you're in the browser or an environment that does not have `window` or `window.history` such as the 
server, React Native or tests (note: Jest does have a fake functioning `window` object, so in Jest tests, you should
use `createMemoryHistory` to keep tests isolated).

When using `createMemoryHistory` the key is to to provide the initial path as the value (within an array) for `initialEntries`. On the
server this is easy because you can get it from your `request` object such as when using *express*. In tests, you can set it to whatever
you want to trick **Redux First Router** into thinking the app is starting on whatever route you want. In React Native, you get it via
the `Linking` API like this:

```javascript
import { connectRoutes } from 'redux-first-router'
import createHistory from 'history/createMemoryHistory'
import { Linking } from 'react-native'
import config from '../config'

const env = 'development' // should dynamically come from environment variables

const url = await Linking.getInitialURL()
const delimiter = config(env).URI_PREFIX || '://'
const path = url ? `/${url.split(delimiter)[1]}` : '/'
const history = createHistory({ initialEntries: [path] })

const { middleware, enhancer, reducer } = connectRoutes(history)
```


## RoutesMap

The `routesMap` was pretty much covered in the [readme](../README.md), but to be thorough,
we'll explain it in depth, as well as describe the missing details about the `toPath` and `fromPath` functions. Here's its 
*Flow type*:

```javascript
type RoutesMap = {
  [key: string]: string | RouteObject,
}

type RouteObject = {
  path: string,
  capitalizedWords?: boolean,
  toPath?: (value: string, key?: string) => string,
  fromPath?: (pathSegment: string, key?: string) => string,
  thunk?: (dispatch: Function, getState: Function) => Promise<any>,
}
```

*note: one very important thing to note about the resulting dispatched actions is that the payload is expected to always be an object.
When using* **Redux First Router**, *do not dispatch payloads that are primitives such as `number` or `string`.*

Features:
* **route as a string** is simply a path to match to an action type without any transformations
* **capitalizedWords** when true will break apart hyphenated paths into words, each with the first character capitalizedWords
* **toPath** will one-by-one take the keys and values of your payload object and transform them into path segments. So for a payload 
with multiple key/value pairs, it will call `toPath` multiple times, passing in the individual value as the first argument
and the individual key name as the second argument.
* **fromPath** will do the inverse, taking each dynamic path *:segment* and its name (in this case "segment") and pass it to 
`fromPath` multiple times. The first argument is the segment and the second its name as delinated in your `routesMap` object
after colons.
* **thunk** is a function just like what you dispatch when using the `redux-thunk` middleware, taking `dispatch` and `getState`
arguments. NOTE: you do NOT need `redux-thunk` for this to work. On the client, the thunk will be called any time the middleware
detects a matching route. However to properly manage server-side rendering, there are 2 optimizations: 1) on first load on the client *if
server side rendering is detected*, it will not be called because it will be assumed to have been handled on the server and the 
`initialState` on the client hydrated from that. 2) on the server, on first load, it also WILL NOT be called because it is expected
to be handled manually in order to allow you to syncronously `await` its result before sending your HTML to the client. See the
[server side rendering](./docs/server-rendering.md) doc for the idiomatic way to do this.


## Options
Lastly, let's talk about the `options` you can provide. There are 6. Here's its flow type:

```javascript
type Options = {
  location?: string, // default: 'location'
  title?: string,    // default: 'title'
  scrollTop?: boolean,
  restoreScroll?: ((PrevLocationState, LocationState) => boolean | string | array) => ScrollBehavior,
  onBeforeChange?: (Dispatch, GetState) => void,
  onAfterChange?: (Dispatch, GetState) => void,
  onBackNext?: (Dispatch, GetState, HistoryLocation, Action) => void
}
```

* **location** - the `location` lets you specify what key **Redux First Router** should expect its reducer to be attached to in your Redux state tree. 

* **title** - the `title` is similarly the name of the state key for your page title. **Redux First Router** will change your page 
title for you when the route changes, e.g. `document.title = 'foo'`.

* **notFoundPath** - the `notFoundPath` will be set as the `pathname` in actions when the path corresponding to an action cannot be determined. It's worth noting that if path was pushed from the Address Bar, but there was no matching route, we preserve that path and leave it up to you to decide what to do with it (similar to how you can land on incorrect Github pages where nothing exists, but where the URL stays the same). There are a few other cases where we may know the path for sure, but can't successfull fulfill the request. The path is preserved *(and `notFoundPath` is NOT used)* when we can.

* **scrollTop** - the `scrollTop` option calls `window.scrollTo(0, 0)` on route changes so the user starts each page at the top. This is a *"poor man's"* scroll
restoration, and should be fine while developing, especially if you're using Chrome. Though hash links won't fully function. See the next option for full-on scroll restoration support.

* **restoreScroll** - the `restoreScroll` is a call to `redux-first-router-restore-scroll`'s restoreScroll` function, with a `shouldUpdateScroll` callback passed a single argument. See the [scroll restoration doc](./scroll-restoration.md) for more info. 

* **onBeforeChange** - `onBeforeChange` is a simple function that will be called before the routes change. It's passed your standard `dispatch` and `getState` arguments
like a thunk.

* **onAfterChange** - `onAfterChange` is a simple function that will be called after the routes change. It's passed your standard `dispatch` and `getState` arguments
like a thunk.

* **onBackNext** - `onBackNext` is a simple function that will be called whenever the user uses the browser *back/next* buttons. It's passed your standard `dispatch` and `getState` arguments like a thunk. Actions with kinds `back`, `next`, and `pop` trigger this.

* **navigators** - `navigators` is a map of of your Redux state keys to *React Navigation* navigators. Here's how you do it:


```js
import reduxNavigation from 'redux-first-router-navigation'

const options = {
  navigators: reduxNavigation({
    myStack: MyStackNavigator
  })
}
```

> See the [Redux Navigation](./redux-navigation) docs for info on how to use *React Navigation* with this package. We think you're gonna love it :)
