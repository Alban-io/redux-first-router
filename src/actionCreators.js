export const INIT = '@@address-bar/INIT' //used internally by middleware and then action is re-written
export const NOT_FOUND = '@@address-bar/NOT_FOUND'


/** EXPORTED INIT ACTION CREATORS: 
 *  for use when ready key is not provided and 
 *  manual control of initialization is desired
*/

export function init(pathname) {
  return {
    type: INIT,
    payload: {
      pathname,
    }
  }
}

export function initThunk() {
  return (dispatch, getState) => {
    let {pathname} = getState().location
    return dispatch(init(pathname))
  }
}


/** PRIMARY/SOLE NAVIGATION ACTION CREATOR 
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'