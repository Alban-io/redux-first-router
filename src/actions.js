export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

/** PRIMARY/SOLE NAVIGATION ACTION CREATOR
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'
