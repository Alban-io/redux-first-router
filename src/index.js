export {
  default as connectRoutes,
  go,
  back,
} from './connectRoutes'

export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

export { default as actionToPath } from './pure-utils/actionToPath'
export { default as pathToAction } from './pure-utils/pathToAction'

export type {
  RouteString,
  RouteObject,
  Route,
  RoutesMap,
  Routes,
  RouteNames,
  Options,
  Params,
  Payload,
  LocationState,
  Location,
  Meta,
  Action,
  ReceivedAction,
  Listener,
  Listen,
  Push,
  GoBack,
  History,
  HistoryLocation,
  Document,
 } from './flow-types'

