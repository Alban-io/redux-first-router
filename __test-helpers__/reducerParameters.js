import createSmartHistory from '../src/smart-history'

import { createInitialState } from '../src/createLocationReducer'

export default async (type, pathname) => {
  // eslint-disable-line import/prefer-default-export
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  await history.push(pathname)

  const current = { pathname, type, payload: { param: 'bar' } }
  const prev = { pathname: '/first', type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: { path: '/first' },
    SECOND: { path: '/second/:param' }
  }

  return {
    type,
    pathname,
    current,
    prev,

    initialState: createInitialState(
      routesMap,
      history
    ),

    routesMap,
    history,

    action: {
      type,
      payload: { param: 'bar' },
      meta: {
        location: {
          current,
          prev,
          kind: 'load',
          history: {
            entries: history.entries.slice(0), // history.entries.map(entry => entry.pathname)
            index: history.index,
            length: history.length
          }
        }
      }
    },

    expectState(state) {
      expect(state.pathname).toEqual(pathname)
      expect(state.type).toEqual(type)
      expect(state.payload).toEqual({ param: 'bar' })
      expect(state.prev).toEqual(prev)
      expect(state.kind).toEqual('load')

      expect(state).toMatchSnapshot()
    }
  }
}
