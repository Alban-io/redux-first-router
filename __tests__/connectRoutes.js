import { createStore, applyMiddleware } from 'redux'

import setup from '../__test-helpers__/setup'
import setupThunk from '../__test-helpers__/setupThunk'
import reducerParameters from '../__test-helpers__/reducerParameters'

import connectRoutes from '../src/connectRoutes'
import { NOT_FOUND } from '../src/index'


describe('middleware', () => {
  it('dispatches location-aware action, changes address bar + document.title', () => {
    const { middleware, history, reducer: locationReducer } = setup()
    const middlewares = applyMiddleware(middleware)

    const reducer = (state = {}, action = {}) => ({
      location: locationReducer(state.location, action),
      title: `title: ${action.type}`,
    })

    const store = createStore(reducer, middlewares)

    console.log(store.getState())

    expect(document.title).toEqual('')
    expect(history.location.pathname).toEqual('/')
    expect(store.getState().location).toMatchObject({
      pathname: '/',
      type: NOT_FOUND,
      payload: {},
    })

    const payload = { param: 'bar' }
    const action = store.dispatch({ type: 'SECOND', payload })

    console.log(action)
    console.log(action.meta)
    console.log(store.getState())

    expect(action).toMatchObject({
      type: 'SECOND',
      payload,
      meta: {
        location: {
          current: { pathname: '/second/bar', type: 'SECOND', payload },
          prev: { pathname: '', type: '', payload: {} },
          load: undefined,
          backNext: undefined,
        },
      },
    })

    expect(document.title).toEqual('title: SECOND')
    expect(history.location.pathname).toEqual('/second/bar')
    expect(store.getState()).toMatchObject({
      location: {
        pathname: '/second/bar',
        type: 'SECOND',
        payload,
        prev: { pathname: '', type: '', payload: {} },
        load: undefined,
        backNext: undefined,
        routesMap: { FIRST: '/first', SECOND: '/second/:param' },
      },
      title: 'title: SECOND',
    })

    expect(action).toMatchSnapshot()
    expect(store.getState()).toMatchSnapshot()
  })

  it('not matched received action dispatches the action as normal with no changes', () => {
    const { middleware, history, reducer: locationReducer } = setup('/first')
    const middlewares = applyMiddleware(middleware)
    const reducer = (state = {}, action = {}) => ({
      location: locationReducer(state.location, action),
      title: action.type,
    })

    const store = createStore(reducer, middlewares)

    console.log(history.location)
    console.log(store.getState())

    expect(history.location.pathname).toEqual('/first')
    expect(store.getState().location).toMatchObject({
      type: 'FIRST',
      pathname: '/first',
      payload: {},
    })

    const beforeState = store.getState().location
    const action = store.dispatch({ type: 'BLA' })
    const afterState = store.getState().location

    console.log(action)
    console.log(history.location)
    console.log(store.getState())

    expect(action).toEqual({ type: 'BLA' })             // final action returned from middleware is the same as initially dispatched
    expect(history.location.pathname).toEqual('/first') // window.location has not changed because action not matched
    expect(store.getState().location).toMatchObject({   // location state has not changed because action not matched
      type: 'FIRST',
      pathname: '/first',
      payload: {},
    })

    expect(afterState).toEqual(beforeState)
  })

  it('user dispatches NOT_FOUND and middleware adds missing info to action', () => {
    const { middleware, reducer: locationReducer } = setup()
    const middlewares = applyMiddleware(middleware)

    const reducer = (state = {}, action = {}) => ({
      location: locationReducer(state.location, action),
      title: action.type,
    })

    const store = createStore(reducer, middlewares)
    const action = store.dispatch({ type: NOT_FOUND })

    console.log(action)
    console.log(action.meta)
    console.log(store.getState())

    expect(action).toMatchObject({
      type: '@@pure-redux-router/NOT_FOUND',
      payload: {},
      meta: {
        location: {
          current: { pathname: '/', type: '@@pure-redux-router/NOT_FOUND', payload: {} },
          prev: { pathname: '', type: '', payload: {} },
          load: undefined,
          backNext: undefined,
        },
      },
    })

    expect(action).toMatchSnapshot()
  })

  it('does nothing and warns if action has error && dispatched action isLocationAction', () => {
    const { middleware, reducer: locationReducer } = setup()
    const middlewares = applyMiddleware(middleware)

    const reducer = (state = {}, action = {}) => ({
      location: locationReducer(state.location, action),
      title: action.type,
    })

    console.warn = jest.fn()
    const store = createStore(reducer, middlewares)
    const receivedAction = { error: true, type: 'FOO', meta: { location: {} } }
    const action = store.dispatch(receivedAction)
    const warnArg = console.warn.mock.calls[0][0]
    expect(warnArg).toEqual('pure-redux-router: location update did not dispatch as your action has an error.')

    console.log(action)
    expect(action).toEqual(receivedAction)
  })
})


describe('middleware -> _middlewareAttemptChangeUrl()', () => {
  it('when pathname changes push new pathname on to addressbar', () => {
    const { _middlewareAttemptChangeUrl } = setup('/')
    const locationState = { pathname: '/foo' }
    const history = []

    _middlewareAttemptChangeUrl(locationState, history)

    console.log(history)
    expect(history).toEqual([{ pathname: '/foo' }])
  })

  it('when pathname does not change, do not push pathname on to address bar', () => {
    const { _middlewareAttemptChangeUrl } = setup('/foo')
    const locationState = { pathname: '/foo' }
    const history = []

    _middlewareAttemptChangeUrl(locationState, history)

    console.log(history)
    expect(history).toEqual([])
  })
})


describe('enhancer', () => {
  it('dispatches location-aware action when store is first created so app is location aware on load', () => {
    const { enhancer, reducer: locationReducer } = setup('/first')

    const createStore = (reducer /* , initialState, enhancer */) => ({ // eslint-disable-line arrow-parens
      dispatch: jest.fn(),
      getState: () => reducer(),
    })

    const reducer = (state = {}, action = {}) => ({
      location: locationReducer(state.location, action),
    })

    const store = enhancer(createStore)(reducer)
    const action = store.dispatch.mock.calls[0][0]

    console.log(action)

    expect(action).toMatchObject({
      type: 'FIRST',
      payload: {},
      meta: {
        location: {
          current: {
            type: 'FIRST',
            payload: {},
            pathname: '/first',
          },
          load: true, // IMPORTANT: only dispatched on load
        },
      },
    })

    expect(action).toMatchSnapshot()
  })

  it('listens to history changes and dispatches actions matching history\'s location.pathname', () => {
    const { history, enhancer, reducer: locationReducer } = setup('/first')

    const createStore = (reducer /* , initialState, enhancer */) => ({ // eslint-disable-line arrow-parens
      dispatch: jest.fn(),
      getState: () => reducer(),
    })

    const reducer = (state = {}, action = {}) => ({
      location: locationReducer(state.location, action),
    })

    const store = enhancer(createStore)(reducer)

    history.push('/second/bar')

    let action = store.dispatch.mock.calls[1][0]
    console.log(action)

    expect(action).toMatchObject({
      type: 'SECOND',
      payload: { param: 'bar' },
      meta: {
        location: {
          current: {
            type: 'SECOND',
            payload: { param: 'bar' },
            pathname: '/second/bar',
          },
          prev: {
            type: 'FIRST',
            payload: {},
            pathname: '/first',
          },
          backNext: true,  // IMPORTANT: only dispatched when using browser back/forward buttons
        },
      },
    })

    expect(action).toMatchSnapshot()

    history.goBack()
    action = store.dispatch.mock.calls[2][0]

    console.log(action)
    expect(action.type).toEqual('FIRST')
    expect(action.meta.location.current.pathname).toEqual('/first')
  })

  it('throws when no location reducer provided', () => {
    const { enhancer, reducer: locationReducer } = setup('/first')

    const createStore = (reducer /* , initialState, enhancer */) => ({ // eslint-disable-line arrow-parens
      dispatch: jest.fn(),
      getState: () => reducer(),
    })

    const reducer = (state = {}, action = {}) => ({
      locationFOO: locationReducer(state.location, action),
    })

    const createEnhancer = () => enhancer(createStore)(reducer)
    expect(createEnhancer).toThrowError()
  })

  it('on the client correctly assigns routesMap to preloadedState so that functions in stringified server state are put back', () => {
    const { enhancer, reducer: locationReducer } = setup('/first')

    const createStore = (reducer /* , initialState, enhancer */) => ({ // eslint-disable-line arrow-parens
      dispatch: jest.fn(),
      getState: () => reducer(),
    })

    const reducer = (state = {}, action = {}) => ({
      location: locationReducer(state.location, action),
    })

    const preloadedState = { location: {} }
    enhancer(createStore)(reducer, preloadedState)

    console.log(preloadedState)
    expect(preloadedState.location.routesMap).toBeDefined()
  })
})


describe('enhancer -> _historyAttemptDispatchAction()', () => {
  it('dispatches action matching pathname when history location changes', () => {
    const dispatch = jest.fn()
    const historyLocation = { pathname: '/second/foo' }
    const { _historyAttemptDispatchAction } = setup()

    _historyAttemptDispatchAction(dispatch, historyLocation)

    const action = dispatch.mock.calls[0][0]
    console.log(action)

    expect(action.type).toEqual('SECOND')
    expect(action.payload).toEqual({ param: 'foo' })
    expect(action.meta.location.backNext).toEqual(true)

    expect(action).toMatchSnapshot()
  })

  it('does not dispatch if pathname is the same (i.e. was handled by middleware already)', () => {
    const dispatch = jest.fn()
    const historyLocation = { pathname: '/second/foo' }
    const { _historyAttemptDispatchAction } = setup()

    _historyAttemptDispatchAction(jest.fn(), historyLocation)
    _historyAttemptDispatchAction(dispatch, historyLocation)

    // insure multiple dispatches are prevented for the same action/pathname
    // so that middleware and history listener don't double dispatch
    console.log(dispatch.mock.calls)
    expect(dispatch.mock.calls).toEqual([])
  })

  it('calls onBackNext handler with location object on path change', () => {
    const dispatch = jest.fn()
    const historyLocation = { pathname: '/second/foo' }
    const onBackNext = jest.fn()
    const { _historyAttemptDispatchAction } = setup('/', { onBackNext })

    _historyAttemptDispatchAction(dispatch, historyLocation)

    const args = onBackNext.mock.calls[0]
    const action = args[0]
    const histLocation = args[1]

    console.log(action)
    console.log(histLocation)

    expect(action.meta.location.current.pathname).toEqual('/second/foo')
    expect(histLocation.pathname).toEqual('/second/foo')
  })
})


describe('thunk', () => {
  it('middleware:attemptCallRouteThunk DOES calls thunk if locationState.load && !locationState.hasSSR', () => {
    const thunk = jest.fn()
    setupThunk('/second/bar', thunk)

    console.log(thunk.mock.calls[0])

    expect(thunk.mock.calls).not.toEqual([])
    expect(thunk.mock.calls.length).toEqual(1)
    expect(thunk.mock.calls[0].length).toEqual(2)
  })

  it('middleware:attemptCallRouteThunk does NOT call thunk if locationState.load && locationState.hasSSR', () => {
    const thunk = jest.fn()

    global.window.SSRtest = true
    const { store } = setupThunk('/second/bar', thunk)
    delete global.window.SSRtest

    console.log(store.getState())
    console.log(thunk.mock.calls)

    expect(thunk.mock.calls).toEqual([])
  })

  it('middleware:attemptCallRouteThunk DOES call thunk if !locationState.load', () => {
    const thunk = jest.fn()
    const { store } = setupThunk('/first', thunk)

    const action = { type: 'SECOND', payload: { param: 'bar' } }
    store.dispatch(action)

    console.log(store.getState())
    console.log(thunk.mock.calls[0])

    expect(thunk.mock.calls).not.toEqual([])
    expect(thunk.mock.calls[0].length).toEqual(2)
  })

  it('attemptCallRouteThunk calls thunk with same `dispatch` argument as in middleware chain', () => {
    const thunk = jest.fn((dispatch, getState) => {
      const action = { type: 'THIRD', payload: { param: 'hurray' } }
      dispatch(action)

      return getState()
    })

    const { store } = setupThunk('/first', thunk)

    // thunk will be called
    const action = { type: 'SECOND', payload: { param: 'bar' } }
    store.dispatch(action)

    const { location } = store.getState()
    console.log(location)

    // expect state matched that was dispatched in thunk
    expect(location.type).toEqual('THIRD')
    expect(location.pathname).toEqual('/third/hurray')
  })

  it('simulate server side manual usage of thunk via `await connectRoutes().thunk` (to be used when: locationState.load && locationState.hasSSR)', async () => {
    const thunk = jest.fn(async (dispatch, getState) => {
      const action = { type: 'THIRD', payload: { param: 'hurray' } }
      dispatch(action)

      return await Promise.resolve(getState()) // not really needed, but it's important we are testing what a thunk returning a promise looks like
    })

    global.window.SSRtest = true
    const { store, thunk: ssrThunk } = setupThunk('/second/bar', thunk)
    delete global.window.SSRtest

    // verify thunk has not been called yet
    const { location } = store.getState()
    console.log(location)
    expect(location.pathname).toEqual('/second/bar')

    // verify our manual calling of thunk via `await` has been called to simulate server side usage
    const state = await ssrThunk(store)
    console.log(state)

    expect(state.location.pathname).toEqual('/third/hurray')
    expect(state.location.type).toEqual('THIRD')
    expect(state).toEqual(store.getState())
  })

  it('verify server side call of `await connectRoutes().thunk` returns a promise even if route does not have a thunk', async () => {
    global.window.SSRtest = true
    const { store, thunk: ssrThunk } = setupThunk('/first')
    delete global.window.SSRtest

    // promise resolving to undefined returned
    const ret = await ssrThunk(store)
    console.log(ret)
    expect(ret).not.toBeDefined()
  })
})


describe('reducer', () => {
  it('reducer EXISTS and works (see __tests__/createLocationReducer for all its tests)', () => {
    const { reducer } = setup()
    const {
      action,
      expectState,
    } = reducerParameters('SECOND', '/second/bar')

    const state = reducer(undefined, action)

    console.log(state)
    expectState(state)
  })
})


it('_exportedGo EXISTS and works (see __tests__/clientOnlyApi for `go` function)', () => {
  const { _exportedGo } = setup()
  const action = _exportedGo('/first')

  console.log(action)
  expect(action).toEqual({ type: 'FIRST', payload: {} })
})


it('connectRoutes(undefined): throw error if no history provided', () => {
  expect(() => connectRoutes()).toThrowError()
})
