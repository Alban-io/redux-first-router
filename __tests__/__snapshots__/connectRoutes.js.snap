exports[`enhancer -> _historyAttemptDispatchAction() dispatches action matching pathname when history location changes 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "backNext": true,
      "current": Object {
        "pathname": "/second/foo",
        "payload": Object {
          "param": "foo",
        },
        "type": "SECOND",
      },
      "history": Object {
        "entries": Array [
          "/",
        ],
        "index": 0,
        "length": 1,
      },
      "load": undefined,
      "prev": Object {
        "pathname": "",
        "payload": Object {},
        "type": "",
      },
    },
  },
  "payload": Object {
    "param": "foo",
  },
  "type": "SECOND",
}
`;

exports[`enhancer dispatches location-aware action when store is first created so app is location aware on load 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "backNext": undefined,
      "current": Object {
        "pathname": "/first",
        "payload": Object {},
        "type": "FIRST",
      },
      "history": Object {
        "entries": Array [
          "/first",
        ],
        "index": 0,
        "length": 1,
      },
      "load": true,
      "prev": Object {
        "pathname": "",
        "payload": Object {},
        "type": "",
      },
    },
  },
  "payload": Object {},
  "type": "FIRST",
}
`;

exports[`enhancer listens to history changes and dispatches actions matching history's location.pathname 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "backNext": true,
      "current": Object {
        "pathname": "/second/bar",
        "payload": Object {
          "param": "bar",
        },
        "type": "SECOND",
      },
      "history": Object {
        "entries": Array [
          "/first",
          "/second/bar",
        ],
        "index": 1,
        "length": 2,
      },
      "load": undefined,
      "prev": Object {
        "pathname": "/first",
        "payload": Object {},
        "type": "FIRST",
      },
    },
  },
  "payload": Object {
    "param": "bar",
  },
  "type": "SECOND",
}
`;

exports[`middleware dispatches location-aware action, changes address bar + document.title 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "backNext": undefined,
      "current": Object {
        "pathname": "/second/bar",
        "payload": Object {
          "param": "bar",
        },
        "type": "SECOND",
      },
      "history": Object {
        "entries": Array [
          "/",
          "/second/bar",
        ],
        "index": 1,
        "length": 2,
      },
      "load": undefined,
      "prev": Object {
        "pathname": "",
        "payload": Object {},
        "type": "",
      },
    },
  },
  "payload": Object {
    "param": "bar",
  },
  "type": "SECOND",
}
`;

exports[`middleware dispatches location-aware action, changes address bar + document.title 2`] = `
Object {
  "location": Object {
    "backNext": undefined,
    "hasSSR": undefined,
    "history": Object {
      "entries": Array [
        "/",
        "/second/bar",
      ],
      "index": 1,
      "length": 2,
    },
    "load": undefined,
    "pathname": "/second/bar",
    "payload": Object {
      "param": "bar",
    },
    "prev": Object {
      "pathname": "",
      "payload": Object {},
      "type": "",
    },
    "routesMap": Object {
      "FIRST": "/first",
      "SECOND": "/second/:param",
    },
    "type": "SECOND",
  },
  "title": "title: SECOND",
}
`;

exports[`middleware user dispatches NOT_FOUND and middleware adds missing info to action 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "backNext": undefined,
      "current": Object {
        "pathname": "/",
        "payload": Object {},
        "type": "@@pure-redux-router/NOT_FOUND",
      },
      "history": Object {
        "entries": Array [
          "/",
        ],
        "index": 0,
        "length": 1,
      },
      "load": undefined,
      "prev": Object {
        "pathname": "",
        "payload": Object {},
        "type": "",
      },
    },
  },
  "payload": Object {},
  "type": "@@pure-redux-router/NOT_FOUND",
}
`;

exports[`reducer reducer EXISTS and works (see __tests__/createLocationReducer for all its tests) 1`] = `
Object {
  "backNext": true,
  "hasSSR": undefined,
  "history": Object {
    "entries": Array [
      "/second/bar",
    ],
    "index": 0,
    "length": 1,
  },
  "load": true,
  "pathname": "/second/bar",
  "payload": Object {
    "param": "bar",
  },
  "prev": Object {
    "pathname": "/first",
    "payload": Object {},
    "type": "FIRST",
  },
  "routesMap": Object {
    "FIRST": "/first",
    "SECOND": "/second/:param",
  },
  "type": "SECOND",
}
`;
